design/src/rfcs/rfc0023_serialization_and_deserialization.md

```diff
@@ -0,0 +1,90 @@
+<!-- Give your RFC a descriptive name saying what it would accomplish or what feature it defines -->
+RFC: Serialization and Deserialization
+=============
+
+<!-- RFCs start with the "RFC" status and are then either "Implemented" or "Rejected".  -->
+> Status: RFC
+>
+> Applies to: builder structs and other data types that are connected to builder structs
+
+# Overview
+There has been some github issues requesting for the implementation of Serialize/Deserialize traits provided serde crate.  
+At the current state of sdk, user must implement their own solution in order to serialize/deserialize sdk's datatypes.  
+This is a RFC for implementing serde traits.
+
+# Potential Use Cases
+- Testing
+- Reusable conifiguration, infrastructure as a service
+- Logging
+ 
+... and more!
+
+# Implementation
+Add `serde::Serialize` and `serde::Deserialize` to builder datatypes.
+
+# Discussion
+## Enum Representation 
+serde allows programmers to use one of four different tagging ([internal, external, adjacent and untagged](https://serde.rs/enum-representations.html)) when serializing enum.  
+
+I suggest going with `untagged` variant.  
+This way serialized data can be used with aws-cli's file-based input.  
+
+## Backwards Compatability
+- When new fields are added to builder  
+Fields of builder structs are always Option\<T>.
+
+When serde is unable to find a field when deserializing the data, it assigns `None` to the corresponding field.  
+It is highly unlikely that deserialization of data to fail when new fields are introduced to builder types.
+
+- When fields are removed from builder   
+Users will not be able to access that removed fields via builder, however, users should be able to access it via different datatype. (e.g. `serde_json::Map`)
+
+- feature gate for deriving `serde::Serialize` and `serde::Deserialize`
+We are going to implement them behind a feature-gate.  
+The name of the feature gate is going to be `unstable-serde-de` and `unstable-serde-ser`.
+```rust
+#[cfg_attr(feature = "unstable-serde-ser"), derive(serde::Serialize)]
+#[cfg_attr(feature = "unstable-serde-de"), derive(serde::Deserialize)]
+```
+
+This allow users to have better control of their compile time when compared to implementing them both on behind a same feature-gate.
+
+## implementing `serde::Serialize` and `serde::Deserialize` on builder types
+
+I am going to implement serde traits on existing builder types, instead of introducing new data types for serialization/deserialization.
```

Comment:
Instead of framing this RFC as "What I would/will do", could you frame it as "we should do this". In this case, you'll almost certainly be the person implementing this, but the RFC should be a proposal of what "we" (everyone contributing to this project) should do.

Reply:
OK